pil-analyzer/src/type_inference.rs

```diff
@@ -168,23 +230,25 @@ impl TypeChecker {
 
         // Now we check for all symbols that are not declared as a type scheme that they
         // can resolve to a concrete type.
-        for (name, (source_ref, declared_type)) in &self.declared_types {
-            if declared_type.vars.is_empty() {
-                // It is not a type scheme, see if we were able to derive a concrete type.
-                let inferred = self.type_into_substituted(declared_type.ty.clone());
-                if !inferred.is_concrete_type() {
-                    let inferred_scheme = self.to_type_scheme(inferred);
-                    return Err(source_ref.with_error(
-                        format!(
-                            "Could not derive a concrete type for symbol {name}.\nInferred type scheme: {}\n",
-                            format_type_scheme_around_name(
-                                name,
-                                &Some(inferred_scheme),
-                            )
-                        )));
-                }
-            }
-        }
+        self.declared_types
+            .iter()
+            .filter(|(_, declared_type)| declared_type.vars.is_empty())
```

Comment:
I think `vars.is_empty()` is a very common operation on DeclaredType - maybe this could be a function `is_concrete`?